yapapi/executor/services.py

```diff
@@ -294,53 +294,72 @@ def _get_handler(instance: ServiceInstance):
             return handler()
 
     async def _run_instance(self, ctx: WorkContext):
+
         loop = asyncio.get_event_loop()
         instance = ServiceInstance(service=self._service_class(self, ctx))
         self.__instances.append(instance)
 
         logger.info(f"{instance.service} commissioned")
 
         handler = self._get_handler(instance)
-        batch = None
+
+        batch_task: Optional[asyncio.Task] = None
+        signal_task: Optional[asyncio.Task] = None
 
         while handler:
-            try:
-                if batch:
-                    r = yield batch
-                    fr = loop.create_future()
-                    fr.set_result(await r)
-                    batch = await handler.asend(fr)
-                else:
-                    batch = await handler.__anext__()
-            except StopAsyncIteration:
-                instance.service_state.lifecycle()
-                handler = self._get_handler(instance)
-                batch = None
-                logger.debug(f"{instance.service} state changed to {instance.state.value}")
-
-            # TODO
-            #
-            # two potential issues:
-            # * awaiting a batch makes us lose an ability to interpret a signal (await on generator won't return)
-            # * we may be losing a `batch` when we act on the control signal
-            #
-            # potential solution:
-            # * use `aiostream.stream.merge`
-
-            ctl = instance.get_control_signal()
-            if ctl == ControlSignal.stop:
-                if instance.state == ServiceState.running:
-                    instance.service_state.stop()
-                else:
-                    instance.service_state.terminate()
-
-                logger.debug(f"{instance.service} state changed to {instance.state.value}")
-
-                handler = self._get_handler(instance)
-                batch = None
+            # Repeatedly wait on one of `(batch_task, signal_task)` to finish.
+            # If it's the first one, retrieve a batch from its result and handle it.
+            # If it's the second -- retrieve and handle a signal.
+            # Any finished task is replaced with a new one, so there are always two.
+
+            if batch_task is None:
+                batch_task = loop.create_task(handler.__anext__())
+            if signal_task is None:
+                signal_task = loop.create_task(instance.control_queue.get())
```

Comment:
maybe we should remove `ServiceInstance.get_control_signal` then ?

Reply:
removed in https://github.com/golemfactory/yapapi/pull/408/commits/648c114eb28c5db3ad445b8cb837105645d75a8a